sys/kern/event.c

```diff
@@ -1,15 +1,452 @@
 #include <sys/event.h>
+#include <sys/mimiker.h>
+#include <sys/libkern.h>
+#include <sys/file.h>
+#include <sys/filedesc.h>
+#include <sys/condvar.h>
 #include <sys/errno.h>
+#include <sys/malloc.h>
+#include <sys/device.h>
+#include <sys/proc.h>
+#include <sys/pool.h>
+#include <sys/time.h>
 
+#define KN_HASHSIZE 8
+
+static POOL_DEFINE(P_KNOTE, "knote", sizeof(knote_t));
+
+static kqueue_t *kqueue_create(void);
+static void kqueue_drain(kqueue_t *kq);
+static void kqueue_destroy(kqueue_t *kq);
+
+static void knote_enqueue(knote_t *kn);
+static void knote_dequeue(knote_t *kn);
+static void knote_drop(knote_t *kn);
+
+typedef TAILQ_HEAD(, knote) knote_tailq_t;
+
+/* kqueue stands for the kernel event queue. Each event that can be triggered
+ * in a kqueue is represented by knote (owned by that instance of kqueue).
+ *
+ * All fields are protected by `kq_lock`.
+ *
+ * `kn_objlock` (the lock protecting the object) must be taken BEFORE `kq_lock`.
+ */
+typedef struct kqueue {
+  int kq_count;          /* number of pending events */
+  knote_tailq_t kq_head; /* list of pending events */
+  mtx_t kq_lock;         /* mutex for queue access */
+  condvar_t kq_cv;
+  knlist_t kq_knhash[KN_HASHSIZE]; /* hash table for knotes */
+} kqueue_t;
+
+/* Returns a hash bucket for a given object */
+static inline knlist_t *kq_gethash(kqueue_t *kq, void *obj) {
+  return &kq->kq_knhash[((unsigned long)obj) % KN_HASHSIZE];
+}
+
+static kqueue_t *kqueue_create(void) {
+  kqueue_t *kq = kmalloc(M_DEV, sizeof(kqueue_t), M_ZERO);
+  mtx_init(&kq->kq_lock, 0);
+  cv_init(&kq->kq_cv, 0);
+
+  TAILQ_INIT(&kq->kq_head);
+  for (int i = 0; i < KN_HASHSIZE; i++)
+    SLIST_INIT(&kq->kq_knhash[i]);
+
+  return kq;
+}
+
+/* Disconnects all knotes from the given kqueue. */
+static void kqueue_drain(kqueue_t *kq) {
+  knote_t *kn;
+
+  for (int i = 0; i < KN_HASHSIZE; i++) {
+    while ((kn = SLIST_FIRST(&kq->kq_knhash[i])) != NULL) {
+      knote_drop(kn);
+    }
+  }
+}
+
+static void kqueue_destroy(kqueue_t *kq) {
+  kqueue_drain(kq);
+  mtx_destroy(&kq->kq_lock);
+  cv_destroy(&kq->kq_cv);
```

Comment:
NIT: I'm not sure if mtx should be destroyed before condvar that uses this lock.